clients/roscpp/src/libros/spinner.cpp

```diff
@@ -30,10 +30,90 @@
 #include "ros/callback_queue.h"
 
 #include <boost/thread/thread.hpp>
-#include <boost/thread/recursive_mutex.hpp>
+#include <boost/thread/mutex.hpp>
 
 namespace {
-  boost::recursive_mutex spinmutex;
+
+/** class to monitor running single-threaded spinners.
+ *
+ *  Calling the callbacks of a callback queue _in order_, requires a unique SingleThreadedSpinner
+ *  spinning on the queue. Other threads accessing the callback queue will probably intercept execution order.
+
+ *  To avoid multiple SingleThreadedSpinners (started from different threads) to operate on the same callback queue,
+ *  this class stores a map of all spinned callback queues.
+ *  If the spinner is single threaded, the corresponding thread-id is stored in the map
+ *  and if other threads will try to spin the same queue, an error message is issued.
+ *
+ *  If the spinner is multi-threaded, the stored thread-id is NULL and future SingleThreadedSpinners
+ *  should not spin this queue.
+ */
+struct SpinnerMonitor
+{
+  /* store spinner information per callback queue:
+     Only as unique single-threaded spinner is allowed and we store its thread id.
+     As multiple multi-threaded spinners are allowed in parallel, their number is counted.
+  */
+  struct Entry
+  {
+    Entry(const boost::thread::id &tid) : tid(tid) {}
+
+    boost::thread::id tid; // thread id of single-threaded spinner
+    unsigned int num_multi_threaded; // number of multi-threaded spinners
```

Comment:
The current patch allows more then one multi threaded spinners to handle the same queue. @tfoote based on your last paragraph it sounds like you are not agreeing that this should be the case? Am I reading this right?

I am not sure if I understand what you are saying correctly. @rhaschke point is that once you choose to use a multi threaded spinner (with N > 1) you can't rely on the order anyway which I would agree to.

Consider the following example where the queue contains the messages `A, B, C`. A single threaded spinner will take them in order and process each item using whatever action (e.g. creating an `A'` from the `A`). The results will for sure have the same order `A', B', C'` since the time the process each message is irrelevant. But for a multi threaded spinner I don't see how that would currently be guaranteed. While it would still pick the messages out of the queue in the same order due to processing them in parallel the result order might be different, e.g. `B', C', A'` if `A` took significantly longer to process then `B` and `C`.

@tfoote Can you please clarify your statement and your opinion of the current state of the patch.